### ISSUE TITLE
install via NPM

### ISSUE THREAD

**Issue by nelix** [literallycanvas/literallycanvas#230]
The package.json is already present, it would be nice if you added tha package to npm.

**Comment by bikegriffith** [comment]
+1 to this.  Even if we had to manually copy images/css, it'd be closer to being usable in an ES6/Webpack/React app (I'm using react-starter-kit right now).  You can actually grab it via the npm github syntax - `npm install literallycanvas/literallycanvas` - but importing it shows some dependencies on `window` that make babel/webpack cry

**Comment by irskep** [comment]
Here is some additional information about why I have not done this. I do not oppose it being done, but I have not done it.

# 1. Release complexity

We currently have 2 release vectors. One is a plain JavaScript file plus assets distributed via tarball. While it "pollutes" the global namespace with its `LC` object and requires manual setup, it is easy to understand.

The other is a Bower repository. When I first added it, it broke constantly, and we had to make a special snowflake *repository* (not branch!) for it so it wouldn't *crash* because *not all our **tags** are semver format*. TAGS!

Adding a third release vector increases the QA surface area for a release. If you look at the changelog, you'll see that I already do a pretty poor job of release QA. NPM would add one more way to break it.

# 2. It doesn't functionally change anything

Except for generating "warnings" in opinionated JS tools (personally I love webpack but it can screw off with respect to babysitting my usage of `window` or which JS files I choose to use), NPMifying this project would have no benefit. It would do exactly the same thing and have exactly the same API.

# 3. I am not personally motivated

I simply don't care enough to spend the time doing this when I could be improving the API and performance instead. :-)

So, there you have it. The only way this will happen is if someone (a) drops a finished pull request in my lap, or (b) sponsors the work.

**Comment by idolize** [comment]
@irskep I'm happy to make this work with npm/CommonJS via PR if you're still open to that.

Npm takes *a lot less* babysitting than Bower. Just update `package.json` and `npm publish`; no fiddling with tags or separate repos or anything like that.

-----

You are right, however, that there is no "one true solution" to the issue of embedding images/CSS via CommonJS. It's not really a limitation of npm, which will happily upload/download whatever non-JS assets you want, but rather the way of accessing them the "right" way in your CommonJS code (`require()` was originally only meant for JS files).

Since you already allow configuration of the path for the image assets, I think that should continue to work fine (the developer will just have to point the path to the `./node_modules` folder, or wherever else the images may be saved - same as it is now).

Including CSS, on the other hand, isn't as obvious. The solutions I've seen range from just getting the developer to copy the CSS themselves (possibly by reaching into `./node_modules` again)–which works fine but isn't quite as elegant–or using `require('./someCSSFile.css')` with Webpack's [CSS loader](https://github.com/webpack/css-loader)/Browserify's [CSSify](https://github.com/davidguttman/cssify) transform.

The second option is the approach taken by libraries such as [react-spinkit](https://github.com/KyleAMathews/react-spinkit#css), and it works great for all users of Webpack **and** Browserify, which I imagine would be the prinary use-case for this library when not using the pre-built `.min.js` files.

**Comment by idolize** [comment]
As far as "NPMifying this project [having] no benefit": I would have to strongly disagree. Npm is rapidly becoming *the* way to include/manage JavaScript libraries. It greatly decreases the burden on developers including/managing a lot of dependencies in their projects. It ensures we get the latest changes when libraries update, makes rolling back versions easier, etc.

I would go so far as to say that using a package manager for dependencies is considered a best practice for any language/platform, and [Bower is definitely on the way out](https://github.com/rackt/redux/issues/944#issuecomment-154858804).

**Comment by irskep** [comment]
I know people have their own workflows, and I have no objection to LC being on NPM. I just don't want to have to manage it myself except to run some command to update it at release time. :-)

**Comment by irskep** [comment]
If someone wants to take the time to set it up, I will happily add them as collaborators to this repo and make sure all relevant parties can update the package on NPM.

**Comment by idolize** [comment]
@irskep Awesome :smile:

(Sorry if I was overly verbose in my response)

I'm happy to set it up- I've done it before with several repos, so it shouldn't be too difficult.

**Comment by irskep** [comment]
OK. From a maintainer's perspective, here's what I would like in addition to it pretty much working:

1. A subdirectory containing a sanity check so I can quickly verify that a release was uploaded correctly
2. A one-line way to push the latest release to NPM, in the Makefile if possible but not necessarily.

**Comment by idolize** [comment]
Got it.

Would you also be open to (in a different PR) updating the React version to the latest API (0.14), making it a [`peerDependency` so it can co-exist with other React components](https://stackoverflow.com/questions/30451556/what-is-the-correct-way-of-adding-a-dependency-to-react-in-your-package-json-for), and making the React components directly accessible? See #355

**Comment by irskep** [comment]
#353 upgraded React to v0.14. 

As for `peerDependencies` - since LC should be used as a widget inside a larger app, it does make sense for React to be a `peerDependency`, so it would be okay to make that change too.

I'll keep #355 discussion on that ticket.

**Comment by idolize** [comment]
FYI #353 did not completely update React to 0.14. React 0.14 requires [another dependency on `react-dom`](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#major-changes).

```
Warning: React.render is deprecated. Please use ReactDOM.render from require('react-dom') instead.
```

**Comment by idolize** [comment]
@irskep What is the current workflow for releasing new versions of literallycanvas?

I'm guessing it is as follows:

```
make
make release-files
make release-branch
make release-bower
```

The only additional build step I want to add is one to compile the coffeescript to JS before publishing to npm (not go so far as to Browserify it together into a bundle for browsers, but rather just [make the coffeescript into CommonJS](http://stackoverflow.com/questions/13645824/can-i-write-npm-package-in-coffeescript)).

Right now you use the `lib/js` directory for Browserified output. I could make another directory for the non-Browserified JS or I could just use `lib/js` again. The downside of using the same directory is that you need to keep the order of operations for your builds correct or you could end up releasing Browserified JS to npm or CommonJS to your `<script>` tag users.

I'd recommend adding a [`prepublish` hook to your `package.json`](https://docs.npmjs.com/misc/scripts) so that all you ever have to do to publish to npm is run "npm publish" and it will always do a `coffee` compile beforehand (that may you can never accidentally forget).

Then the `release-npm` make task becomes just as alias for `npm publish`.

**Comment by irskep** [comment]
Adding to the 0.4.12 milestone since all the work is done except to publish the next release.

**Comment by idolize** [comment]
@irskep When you you think you'd be able to [publish this release to npm](https://docs.npmjs.com/getting-started/publishing-npm-packages)?

Thanks again!

**Comment by irskep** [comment]
I'm shooting for Friday. If people test it thoroughly, could be sooner.

**Comment by idolize** [comment]
Thanks!

FYI - I just tried updating my project from using my fork over over to using the upstream `master`. It went very smoothly once I did a `gulp commonjs` (since the built files aren't checked into GitHub) and I renamed `import { LiterallyCanvas } from 'literallycanvas'` to `import { LiterallyCanvasReactComponent as LiterallyCanvas } from 'literallycanvas'`.

No issues so far in my testing.

**Comment by idolize** [comment]
@irskep Have you had a chance to test?

**Comment by irskep** [comment]
Someone's an eager beaver.

**Comment by idolize** [comment]
Haha yeah, I'd like to avoid depending on a github repo in my `package.json` before I check in code.

**Comment by irskep** [comment]
Right now I am just trying to get a demo project to 'npm install'. Unfortunately npm is total garbage so it hasn't finished yet.

**Comment by irskep** [comment]
This ought to work straightforwardly, right?

```son
{
  "name": "lc-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "dependencies": {
    "literallycanvas": "git://github.com/literallycanvas/literallycanvas.git#master",
    "react": "^0.14.6",
    "react-dom": "^0.14.6"
  },
  "devDependencies": {
    "babel-core": "~6.4.0",
    "babel-preset-es2015": "~6.3.13",
    "babel-preset-react": "~6.3.13",
    "webpack": "^1.12.11",
    "babel-loader": "^6.2.1"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Steve Johnson",
  "license": "Public domain"
}

**Comment by idolize** [comment]
Well, if you add the git repo as a dependency then it probably won't work because the git repo has none of the built files (`'lib/js`) are checked in to master.

Try doing `cd node_modules/literallycanvas && gulp commonjs` then run your demo.

**Comment by idolize** [comment]
Also, one other thing to look out for is make sure you don't get two copies of React installed. I think it should work alright because we have React as a `peerDependency` in literallycanvas, but it *is* in `devDependencies`, so if you `npm install` it directly without [`--production`](https://docs.npmjs.com/cli/install) it may install an extra copy.

**Comment by irskep** [comment]
My problem is more that my 'npm install --dev' has been running for 25 minutes now.

**Comment by irskep** [comment]
Solved it by typing "npm install package@version" one at a time...testing now

**Comment by idolize** [comment]
25 mins?? Never seen anything even close to that before.

**Comment by irskep** [comment]
I've hit it regularly on two separate machines. I suspect it's going into an infinite loop somewhere.

Combine that with some password encoding issues I had a while back, and you can see where my disdain comes from. :-)

**Comment by irskep** [comment]
Here's my new package.json:

```json
{
  "name": "lc-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "dependencies": {
    "literallycanvas": "../literallycanvas",
    "react": "^0.14.6",
    "react-dom": "^0.14.6"
  },
  "devDependencies": {
    "babel-core": "~6.4.0",
    "babel-preset-es2015": "~6.3.13",
    "babel-preset-react": "~6.3.13",
    "webpack": "^1.12.11",
    "babel-loader": "^6.2.1"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Steve Johnson",
  "license": "Public domain"
}
```

which just uses my local copy. But even though the require-able JS files exist in the source directory, they don't end up copied to the `node_modules/literallycanvas/lib/js` folder. Any ideas?

**Comment by irskep** [comment]
Figured it out. There was a `.gitignore` file in there which NPM decided to incorrectly copy into `.npmignore`. yaaaaay

**Comment by irskep** [comment]
The NPM build does not use the preprocessor that separates the "core" build. So I can't release on NPM right now because the behavior of the code is undefined under these conditions.

**Comment by irskep** [comment]
Specifically, this code:

```coffee
  # @ifdef INCLUDE_GUI
  return require('./reactGUI/initDOM')(el, opts)
  # @endif
  # @ifndef INCLUDE_GUI
  return initWithoutGUI(el, opts)
  # @endif
```

should have either the top block OR the bottom block, not both.

It happens to work okay right now, but I'm not comfortable with the situation.

**Comment by irskep** [comment]
This line of the gulpfile needs to find its way into the commonjs target:

```js
    .transform(preprocessify({ INCLUDE_GUI: true }, {includeExtensions: ['.coffee'], type: 'coffee'}))
```

**Comment by irskep** [comment]
tl;dr not going to release until one of these things happens:

1. Someone submits a patch to fix the problem
2. I put in the time to figure out how to add preprocessify to the commonjs target. Might not take a lot of actual time, but there is a lot of activation energy involved. Messing with build processes is not fun.
3. I rewrite the whole build process in webpack. This is about equally likely for me as #2. I really don't like gulp or browserify. Webpack I actually understand

**Comment by irskep** [comment]
Fixed it.

I'm currently working on this repository: https://github.com/literallycanvas/literallycanvas-demos

Since this version adds to the release process, I want to be more methodical about documenting and testing methods of distribution. I just need to add one more demo for the "traditional" plain-js-file method and play with all the tools to make sure they still work. Over time I'll add API examples. But I'll release as soon as I've made that second demo and played around with the UI.

**Comment by irskep** [comment]
I'm now satisfied that the NPM release process will work fine. I just happened to reproduce #359, so I'm fixing that now.

**Comment by idolize** [comment]
:+1:

**Comment by irskep** [comment]
It's up now: https://www.npmjs.com/package/literallycanvas

So I guess this is now irrevocably version 0.4.12, since I can't update anything. That 'state of the project' readme segment isn't accurate anymore but it'll stay up until the next release.

Personally I would use the word "asinine" to describe intentionally disallowing documentation edits on a published version, but what's done is done.

**Comment by idolize** [comment]
Awesome thanks!

They used to allow pushing changes without upping the version number, but that was changed to prevent people editing non-docs files and publishing under the same version number. There's really not much harm in publishing a new patch version with a doc edit.

**Comment by irskep** [comment]
Screw it then, I'm renaming this release to 0.4.13.

**Comment by idolize** [comment]
It's working great on my end!
